config/src/main/java/org/springframework/security/config/oauth2/client/ClientRegistrationsBeanDefinitionParser.java

```diff
@@ -104,8 +105,10 @@ public BeanDefinition parse(Element element, ParserContext parserContext) {
 				}
 			}
 			getOptionalIfNotEmpty(clientRegistrationElt.getAttribute(ATT_CLIENT_ID))
+					.map(parserContext.getReaderContext().getEnvironment()::resolvePlaceholders)
```

Comment:
There may be other placeholders that need to be resolved other than `client-id` and `client-secret` and I would like to avoid adding `parserContext.getReaderContext().getEnvironment()::resolvePlaceholders` to every element.

If you look at this [commit](https://github.com/spring-projects/spring-security/commit/4542f00b14ba6dc8bb8c1cf4365dc7dad5fe5a3c#diff-6a79f880c5e53b19f12658d702e68cbaR325), could we simply register a `org.springframework.beans.factory.config.PropertyPlaceholderConfigurer` `@Bean` to take care of it?

Reply:
Thanks for the review @jgrandja.
In this case, we do not register `ClientRegistration` as a bean, so we need to manually resolve property placeholders.

Maybe it would be better to move `parserContext.getReaderContext().getEnvironment()::resolvePlaceholders` to `getOptionalIfNotEmpty` method?

Reply:
Yes you are right @evgeniycheban - `ClientRegistration` are not beans so that won't work.

> Maybe it would be better to move `parserContext.getReaderContext().getEnvironment()::resolvePlaceholders` to `getOptionalIfNotEmpty` method?

Agreed.